[jslib] Refactored Acknowledgements with Enhanced HTTP Utilities

HenriRabalais opened the pull request:
This PR refactors the acknowledgements page to use a new, more robust and reusable API client and a suite of custom error classes. This change modernizes our data fetching approach by moving away from raw fetch requests and adopting a structured, type-safe methodology.

### Key Changes
Introduced a new Core Library: Added a new directory at jslib/core containing:
 - Client.ts: A generic, extensible class for handling API requests (GET, POST, PUT). It includes robust error handling for network issues, HTTP response statuses (e.g., 404), and JSON parsing errors.
 - Query.ts: A class for building type-safe URL query strings, supporting various operators and parameters.
 - Added Custom Error Classes: A comprehensive set of custom error classes are now available at jslib/core/errors to provide more context for debugging API failures. This includes ApiNetworkError, ApiResponseError, JsonParseError, and more.

Introduced a Entities Library: Added a new directory at jslib/entities containing:
- Acknowledgment entity and affiliated type and client
- This will eventually house all LORIS entities, and contexts they provide, specific errors, hooks, components, etc.

### Updated Acknowledgements Module
- The acknowledgementsIndex.js file has been refactored to use the new AcknowledgementClient instead of a raw fetch call. This significantly cleans up the component's logic and separates concerns.
- The AcknowledgementClient is a specialized client that extends the new generic Client.ts, providing an interface for interacting with the acknowledgements API.

### Misc
- Module and Path Configuration: Updated package.json, tsconfig.json, and webpack.config.ts to support the new libraries, including:
- New dependencies: stream-browserify, pbkdf2, and transliteration.

HenriRabalais commented on jslib/core/http/Client.ts:
```diff
@@ -0,0 +1,105 @@
+export class Client<T> {
+  protected baseUrl: string;
+  protected subEndpoint?: string;
+  public getCustomMessage: (
```
this is to allow each Client concrete class to define it's own errors across all the error types, not sure if it's the best implementation, but it seems to work

HenriRabalais commented on jslib/core/errors/BaseError.ts:
```diff
@@ -0,0 +1,12 @@
+ * Base class for all custom API-related errors.
+ */
+export class BaseError extends Error {
+  public name: string;
```
need to delete this since name is already a property of `Error`

ridz1208 commented:
@skarya22 @marandmart I assigned the both of you here because its a JSX heavy PR but also brongs some nice additions to what we already have. This is not exactly a critical priority but definitely a nice have so if you ca take a bit of time here and there to review it it would be perfect. We always talk baout improving our react side and this is doing just that.

marandmart commented on jslib/entities/acknowledgement/clients/AcknowledgementClient.ts:
```diff
@@ -0,0 +1,14 @@
+import {Acknowledgement} from '../../';
+import {Http} from 'jsx/../jslib/core';
```
If there isn't a specific reason for this import being done like this, change to /jslib/core

marandmart commented on jslib/core/http/Query.ts:
```diff
@@ -0,0 +1,115 @@
+    case Operator.GreaterThan: return '>';
+    case Operator.LessThanOrEqual: return '<=';
+    case Operator.GreaterThanOrEqual: return '>=';
+    case Operator.Like: return '_like';
```
The Operator.Includes case is missing

marandmart commented on jslib/core/http/Client.ts:
```diff
@@ -0,0 +1,154 @@
+  protected async fetchJSON<U>(url: string, options: RequestInit): Promise<U> {
+    const request = new Request(url, options);
+    try {
+      const response = await fetch(url, options);
```
Doesn't really change much, but it looks cleaner to use the request const from line 121 here instead of passing the url and options again

HenriRabalais commented:
@marandmart 

Thank you so much for the comprehensive review! All amazing points and feedback. I've addressed and corrected every point you made.

Could you look it over one more time?

Thanks!

skarya22 commented on modules/acknowledgements/jsx/acknowledgementsIndex.js:
```diff
@@ -103,14 +105,17 @@ class AcknowledgementsIndex extends Component {
+    const client = new Acknowledgement.Client();
+    try {
+      const acknowledgements = await client.get(query);
+      console.log({acknowledgements}, {depth: null, colors: true});
```
Leftover console.log

skarya22 commented on jslib/core/http/Query.ts:
```diff
@@ -0,0 +1,116 @@
+  LessThanOrEqual = '<=',
+  GreaterThanOrEqual = '>=',
+  Like = 'like',
+  Includes = 'in'
```
Below, these are _like and _in. Do these need to be consistent with eachother?

HenriRabalais replied:
hmm yeah, now that I look at it I can likely get rid of the function below!

skarya22 commented on package.json:
```diff
@@ -17,17 +17,20 @@
         "jstat": "^1.9.5",
         "jszip": "^3.10.1",
         "papaparse": "^5.3.0",
+        "pbkdf2": "^3.1.3",
```
I don't love the idea of adding additional dependencies to LORIS, could you try to work around these? Might be helpful to look at the `dataquery` module. Dave was able to set-up streams without importing `stream-browserify`, and the other ones. I can't see why `pbkdf2` and `transliteration` would be needed either.

HenriRabalais commented:
@skarya22 Have I addressed your review?

skarya22 commented:
@HenriRabalais looks like it's failing some tests, but the changes look good! Once you have fixed them can you remove the Needs Work label so it looks ready for review?

kongtiaowang commented on jslib/core/http/Client.ts:
```diff
@@ -0,0 +1,183 @@
+   *
+   * @param baseURL The base URL for the API requests.
+   */
+  constructor(baseURL: string) {
```
+constructor(baseURL: string) {
+  // Reproduce the old behaviour: loris.BaseURL + '/' + baseURL + '/'
+  const origin = window.location.origin;  // always https://...
+  const full = `${origin}/${baseURL}/`;
+  this.baseURL = new URL(full);
+}

HenriRabalais commented on jslib/entities/acknowledgement/clients/AcknowledgementClient.ts:
```diff
@@ -0,0 +1,14 @@
+import {Acknowledgement} from '../../';
```
The intent is to keep all the entities that exist across loris in a centralized location so that their types, clients, and eventually hooks, etc. can be imported in a reliable manner. Acknowledgements is pretty secluded, but the design is meant to eventually incorporate entities that are either used across modules, or don't have their own dedicated module like projects, centers, etc.
